File: Geek Assistant (Android AIO)/Modules/General/Action/GA_SwitchTheme.vb
Imports Transitions

Module GA_SetTheme
    Public Sub Run(FormName As String, Optional initiating As Boolean = False)
        ' DarkTheme = My.Settings.DarkTheme
        initiatingbool = initiating

        'If initiating And Not My.Settings.DarkTheme And FormName <> "Main" Then
        '    Exit Sub
        'End If

        Select Case FormName
            Case AppMode.Name
                SetControlTheme(AppMode)
                AppMode.startup_dontShow.ForeColor = Current_fgColor()
            Case Donate.Name
                Donate_Theme()
            Case info.Name
                info_Theme()
            Case Main.Name

                If My.Settings.VerboseLogging Then
                    Dim ThemeString = "dark theme"
                    If My.Settings.DarkTheme Then ThemeString = "Light Theme"
                    GA_SetProgressText.Run($"Switched to {ThemeString}.", -1)
                End If

                If Not initiating And My.Settings.PerformAnimations Then
                    If My.Settings.DarkTheme Then
                        Main.SwitchTheme_Back_UI.Top = -Main.SwitchTheme_Back_UI.Height
                        'Main.SwitchTheme_Mid_UI.Top = -Main.SwitchTheme_Back_UI.Height - 50
                        'Main.SwitchTheme_Fore_UI.Top = -Main.SwitchTheme_Fore_UI.Height - 100
                        Transition.run(Main.SwitchTheme_Back_UI, "Top", Main.Height, New TransitionType_CriticalDamping(1000))
                        'Transition.run(Main.SwitchTheme_Mid_UI, "Top", Main.Height + 50, New TransitionType_CriticalDamping(1000))
                        'Transition.run(Main.SwitchTheme_Fore_UI, "Top", Main.Height + 100, New TransitionType_CriticalDamping(1000))
                    Else
                        Main.SwitchTheme_Back_UI.Top = Main.Height
                        'Main.SwitchTheme_Mid_UI.Top = Main.Height + 50
                        'Main.SwitchTheme_Fore_UI.Top = Main.Height + 100
                        Transition.run(Main.SwitchTheme_Back_UI, "Top", -Main.SwitchTheme_Back_UI.Height, New TransitionType_CriticalDamping(1000))
                        'Transition.run(Main.SwitchTheme_Mid_UI, "Top", -Main.SwitchTheme_Mid_UI.Height - 50, New TransitionType_CriticalDamping(1000))
                        'Transition.run(Main.SwitchTheme_Fore_UI, "Top", -Main.SwitchTheme_Fore_UI.Height - 100, New TransitionType_CriticalDamping(1000))
                    End If
                End If

                MainTheme_delayTimer.Enabled = True
            Case PleaseWait.Name
                PleaseWait_Theme()
            Case GA_Settings.Name
                Settings_Theme()
            Case ToU.Name
                ToU_Theme()
            Case Preparing.Name
                Preparing.Preparing_Label.ForeColor = Current_fgColor()
                Preparing.BackColor = Current_bgColor()
        End Select
    End Sub


#Region "Donate"

    Private Sub Donate_Theme()
        SetControlsArrTheme({Donate,
                            Donate.BitcoinAddress,
                            Donate.BitcoinNote})
        If My.Settings.DarkTheme Then
            Donate.ButtonsBG_UI.BackColor = Color.FromArgb(32, 32, 32)
            Donate.Close_Button.FlatAppearance.MouseOverBackColor = Color.FromArgb(64, 64, 64)
            Donate.pic.Image = My.Resources.DonateHeart_Dark_64
            Donate.title.ForeColor = Color.FromArgb(255, 191, 217)
            Donate.BitcoinAddressQR.Image = My.Resources.BTCAddressQR_Dark
            Donate.GooglePayLink.Image = My.Resources.linkIcon_dark_16
        Else
            Donate.ButtonsBG_UI.BackColor = Color.WhiteSmoke
            Donate.Close_Button.FlatAppearance.MouseOverBackColor = Color.Silver
            Donate.pic.Image = My.Resources.DonateHeart_64
            Donate.title.ForeColor = Color.FromArgb(128, 0, 87)
            Donate.BitcoinAddressQR.Image = My.Resources.BTCAddressQR
            Donate.GooglePayLink.Image = My.Resources.linkIcon_16
        End If
        Donate.GeekAssistant_PictureBox.BackColor = Donate.ButtonsBG_UI.BackColor
    End Sub

#End Region

#Region "info"
    Private Sub info_Theme()
        SetControlsArrTheme({info,
                             info.info_PictureBox,
                             info.msg_Textbox,
                             info.No_Button})
        If My.Settings.DarkTheme Then
            info.ButtonsBG_UI.BackColor = Color.FromArgb(32, 32, 32)
            info.Copy_Button.FlatAppearance.MouseOverBackColor = Color.FromArgb(64, 64, 64)
            info.Yes_Button.FlatAppearance.MouseOverBackColor = Color.FromArgb(64, 64, 64)
            info.No_Button.FlatAppearance.MouseOverBackColor = Color.FromArgb(64, 64, 64)
        Else
            info.ButtonsBG_UI.BackColor = Color.WhiteSmoke
            info.Copy_Button.FlatAppearance.MouseOverBackColor = Color.Silver
            info.Yes_Button.FlatAppearance.MouseOverBackColor = Color.Silver
            info.No_Button.FlatAppearance.MouseOverBackColor = Color.Silver
        End If
        info.GeekAssistant_PictureBox.BackColor = info.ButtonsBG_UI.BackColor
        'info.Copy_Button.ForeColor = Current_fgColor()
        'info.Copy_Button.BackColor = Current_bgColor()
        'info.Close_Button.ForeColor = Current_fgColor()
        'info.Close_Button.BackColor = Current_bgColor()
        'info.DownloadDrivers_Button.ForeColor = Current_fgColor()
        'info.DownloadDrivers_Button.BackColor = Current_bgColor()


    End Sub
#End Region

#Region "Main"
    Private WithEvents MainTheme_delayTimer As New Timer() With {.Interval = 100}
    Private initiatingbool As Boolean = False
    Private Sub MainTheme_delayTimer_Tick(sender As Object, e As EventArgs) Handles MainTheme_delayTimer.Tick
        Dim Controls_array As Object() = {Main,
                                          Main.ManualInfo_GroupBox,
                                          Main.ProgressFakeBG_UI,
                                          Main.Main_Tabs,
                                          Main.PrepareYourDevice_Tab,
                                          Main.FlashZip_Tab,
                                          Main.MoreTools_Tab,
                                          Main.log,
                                          Main.BootloaderUnlockable_CheckBox,
                                          Main.CustomROM_CheckBox,
                                          Main.CustomRecovery_CheckBox
                                         }
        Dim MetroControls_array As Object() = {
                                               Main.Main_Tabs,
                                               Main.PrepareYourDevice_Tab,
                                               Main.FlashZip_Tab,
                                               Main.MoreTools_Tab,
                                               Main.GA_About_Label,
                                               Main.UnlockBL_Label,
                                               Main.MagiskRoot_Label,
                                               Main.Manufacturer_ComboBox,
                                               Main.AndroidVersion_ComboBox,
                                               Main.ProgressBar,
                                               Main.manualCMD_TextBox,
                                               Main.FlashZip_ChooseFile_TextBox,
                                               Main.UnlockBL_Button,
                                               Main.MagiskRoot_Button,
                                               Main.FlashZip_Button,
                                               Main.ProgressBarLabel
                                              }
        SetControlsArrTheme_Metro(MetroControls_array)


        If initiatingbool Then GoTo NoAnimation
        SetControlsArrTheme(Controls_array)
NoAnimation:
        If Not initiatingbool Then GoTo skipNoAnimation
        For Each c As Object In Controls_array
            c.BackColor = Current_bgColor()
            c.ForeColor = Current_fgColor()
        Next
skipNoAnimation:

        If My.Settings.DarkTheme Then
            'Main.SwitchTheme_Mid_UI.BackColor = Color.FromArgb(235, 245, 235)
            'Transition.run(Main.SwitchTheme_Mid_UI, "BackColor", Color.FromArgb(0, 20, 0), New TransitionType_CriticalDamping(1000))
            Main.Main_ToolTip.BackColor = SystemColors.InfoText
            Main.Main_ToolTip.ForeColor = SystemColors.Info
            With Main
                With .AutoDetectDeviceInfo_Button
                    .ForeColor = Color.FromArgb(95, 191, 119)
                    .BackColor = Color.FromArgb(64, 64, 64)
                    .FlatAppearance.MouseDownBackColor = Color.Honeydew
                    .FlatAppearance.MouseOverBackColor = Color.FromArgb(74, 74, 74)
                    .Image = My.Resources.AutoDetect_dark_64
                End With

                '.DeviceState_Label.ForeColor = Color.FromArgb(95, 191, 119)
                .Donate_Button.Icon = My.Resources.DonateHeart_Dark_24
                .SwitchTheme_Button.Icon = My.Resources.Theme_Dark_24
                .GeekAssistant.Image = My.Resources.Geek_Assistant_Dark
                .Feedback_Button.Icon = My.Resources.Smile_dark_24
                .About_Button.Icon = My.Resources.ToU_dark_24
                .Settings_Button.Icon = My.Resources.Settings_dark_24
                .ShowLog_Button.Icon = My.Resources.Commands_dark_24
                .manualCMD_TextBox.Icon = My.Resources.Commands_dark_16
            End With
        Else 'Light Theme
            'Main.SwitchTheme_Mid_UI.BackColor = Color.FromArgb(0, 20, 0)
            'Transition.run(Main.SwitchTheme_Mid_UI, "BackColor", Color.FromArgb(235, 245, 235), New TransitionType_CriticalDamping(1000))
            Main.Main_ToolTip.BackColor = SystemColors.Info
            Main.Main_ToolTip.ForeColor = SystemColors.InfoText

            With Main
                With .AutoDetectDeviceInfo_Button
                    .ForeColor = Color.FromArgb(0, 128, 32)
                    .BackColor = Color.Honeydew
                    .FlatAppearance.MouseDownBackColor = Color.FromArgb(64, 64, 64)
                    .FlatAppearance.MouseOverBackColor = Color.FromArgb(230, 245, 230)
                    .Image = My.Resources.AutoDetect_64
                End With
                '.DeviceState_Label.ForeColor = Color.FromArgb(0, 128, 32)
                .Donate_Button.Icon = My.Resources.DonateHeart_24
                .SwitchTheme_Button.Icon = My.Resources.Theme_Light_24
                .GeekAssistant.Image = My.Resources.Geek_Assistant
                .Feedback_Button.Icon = My.Resources.Smile_24
                .About_Button.Icon = My.Resources.ToU_24
                .Settings_Button.Icon = My.Resources.Settings_24
                .ShowLog_Button.Icon = My.Resources.Commands_24
                .manualCMD_TextBox.Icon = My.Resources.Commands_16
            End With
        End If
        MainTheme_delayTimer.Enabled = False
    End Sub
#End Region

#Region "PleaseWait"
    Private Sub PleaseWait_Theme()
        SetControlsArrTheme({PleaseWait,
                            PleaseWait.PleaseWait_ProgressText})
        If My.Settings.DarkTheme Then
            PleaseWait.PleaseWait_text.ForeColor = Color.FromArgb(0, 200, 0)
            With PleaseWait.StopProcess_Button
                .ForeColor = Color.FromArgb(230, 255, 230)
                .BackColor = Color.FromArgb(55, 28, 25)
                With .FlatAppearance
                    .MouseOverBackColor = Color.FromArgb(128, 0, 0)
                    .MouseDownBackColor = Color.FromArgb(80, 0, 0)
                End With
            End With
        Else
            PleaseWait.PleaseWait_text.ForeColor = Color.FromArgb(0, 100, 0)
            PleaseWait.StopProcess_Button.BackColor = Color.FromArgb(255, 228, 225)
            With PleaseWait.StopProcess_Button
                .BackColor = Color.FromArgb(255, 228, 225)
                .ForeColor = Color.DarkRed
                With .FlatAppearance
                    .MouseOverBackColor = Color.FromArgb(240, 128, 128)
                    .MouseDownBackColor = Color.FromArgb(128, 0, 0)
                End With
            End With

        End If
    End Sub
#End Region

#Region "Settings"
    Private Sub Settings_Theme()
        SetControlsArrTheme({GA_Settings,
                            GA_Settings.ResetGA_GroupBox,
                            GA_Settings.Close_Button})

        If My.Settings.DarkTheme Then
            GA_Settings.ButtonsBG_UI.BackColor = Color.FromArgb(32, 32, 32)
            GA_Settings.ResetGA_LogsOnly_CheckBox_Label.ForeColor = Color.FromArgb(100, 100, 100)
            GA_Settings.ResetGA_Settings_CheckBox_Label.ForeColor = Color.FromArgb(100, 100, 100)
            GA_Settings.OpenLogsFolder.ForeColor = Color.FromArgb(100, 100, 100)
            GA_Settings.ResetGA.FlatAppearance.MouseOverBackColor = Color.FromArgb(135, 83, 59)
            GA_Settings.ResetGA_SelectAll.Image = My.Resources.SelectAll_W_24
            GA_Settings.ResetGA_SelectAll.FlatAppearance.MouseOverBackColor = Color.FromArgb(64, 64, 64)
            GA_Settings.ResetGA_SelectAll.FlatAppearance.MouseDownBackColor = Color.FromArgb(100, 100, 100)
            GA_Settings.SettingsIcon_PictureBox.Image = My.Resources.Settings_dark_64
            GA_Settings.SettingsTitle_Label.ForeColor = Color.FromArgb(184, 243, 254)
        Else
            GA_Settings.ButtonsBG_UI.BackColor = Color.WhiteSmoke
            GA_Settings.ResetGA_LogsOnly_CheckBox_Label.ForeColor = Color.FromArgb(64, 64, 64)
            GA_Settings.ResetGA_Settings_CheckBox_Label.ForeColor = Color.FromArgb(64, 64, 64)
            GA_Settings.OpenLogsFolder.ForeColor = Color.FromArgb(64, 64, 64)
            GA_Settings.ResetGA.FlatAppearance.MouseOverBackColor = Color.FromArgb(255, 203, 179)
            GA_Settings.ResetGA_SelectAll.Image = My.Resources.SelectAll_B_24
            GA_Settings.ResetGA_SelectAll.FlatAppearance.MouseOverBackColor = Color.Silver
            GA_Settings.ResetGA_SelectAll.FlatAppearance.MouseDownBackColor = Color.FromArgb(64, 64, 64)
            GA_Settings.SettingsIcon_PictureBox.Image = My.Resources.Settings_64
            GA_Settings.SettingsTitle_Label.ForeColor = Color.FromArgb(0, 106, 128)

        End If
        GA_Settings.ResetGA.BackColor = GA_Settings.ButtonsBG_UI.BackColor
        GA_Settings.ResetGA_SelectAll.BackColor = GA_Settings.ButtonsBG_UI.BackColor
        GA_Settings.GeekAssistant_PictureBox.BackColor = GA_Settings.ButtonsBG_UI.BackColor
        GA_Settings.Close_Button.BackColor = GA_Settings.BackColor
    End Sub
#End Region

#Region "ToU"
    Private Sub ToU_Theme()
        SetControlsArrTheme({ToU,
                            ToU.TermsOfUse_Box,
                            ToU.ToU_Reject,
                            ToU.ToU_Accept})

        'Do this hell
        If My.Settings.DarkTheme Then
            ToU.ButtonsBG_UI.BackColor = Color.FromArgb(32, 32, 32)
            ToU.Icon_PictureBox.Image = My.Resources.ToU_dark_64
            ToU.ToUTitle_Label.ForeColor = Color.FromArgb(178, 255, 220)
        Else
            ToU.ButtonsBG_UI.BackColor = Color.WhiteSmoke
            ToU.Icon_PictureBox.Image = My.Resources.ToU_64
            ToU.ToUTitle_Label.ForeColor = Color.FromArgb(0, 102, 71)
        End If

        ToU.Copyright.ForeColor = ToU.ToUTitle_Label.ForeColor
        ToU.AcceptCheck_ToU.ForeColor = Current_fgColor()
        ToU.DontShow_ToU.ForeColor = Current_fgColor()
        ToU.AcceptCheck_ToU.BackColor = ToU.ButtonsBG_UI.BackColor
        ToU.DontShow_ToU.BackColor = ToU.ButtonsBG_UI.BackColor
        ToU.GeekAssistant_PictureBox.BackColor = ToU.ButtonsBG_UI.BackColor

    End Sub
#End Region

#Region "Theming Mechanism"

    Private Sub SetControlsArrTheme(controlsObj As Object())
        For Each controlObj As Object In controlsObj
            SetControlTheme(controlObj)
        Next
    End Sub

    Private Sub SetControlsArrTheme_Metro(controlsObj As Object())
        For Each c As Object In controlsObj
            SetControlTheme_Metro(c)
        Next
    End Sub

    Private Sub SetControlTheme(ControlObj As Object)
        If Not My.Settings.PerformAnimations Then
            ControlObj.BackColor = Current_bgColor()
            ControlObj.ForeColor = Current_fgColor()
        Else
            Transition.run(ControlObj, "BackColor", Current_bgColor(), New TransitionType_CriticalDamping(500))
            Transition.run(ControlObj, "ForeColor", Current_fgColor(), New TransitionType_CriticalDamping(500))

        End If
    End Sub

    Private Sub SetControlTheme_Metro(ControlObj As Object)
        ' Cannot animate "Theme" 'Transition.run(ControlObj, "Theme", Current_Theme_Metro(), New TransitionType_CriticalDamping(500))
        ControlObj.Theme = Current_Theme_Metro()
    End Sub

    Public Function Current_bgColor() As Color
        If My.Settings.DarkTheme Then
            Return bgDark
        Else Return bgLight
        End If
    End Function

    Public Function Current_fgColor() As Color
        If My.Settings.DarkTheme Then
            Return fgDark
        Else Return fgLight
        End If
    End Function

    Public Function Current_Theme_Metro() As MetroFramework.MetroThemeStyle
        If My.Settings.DarkTheme Then
            Return MetroFramework.MetroThemeStyle.Dark
        Else Return MetroFramework.MetroThemeStyle.Default
        End If
    End Function

#Region "Theme Colors"
    Private bgLight As Color = Color.White
    Private fgLight As Color = SystemColors.ControlText
    Private bgDark As Color = Color.FromArgb(17, 17, 17)
    Private fgDark As Color = Color.White
#End Region

#End Region
End Module

File: Geek Assistant (Android AIO)/Modules/General/Action/GA_SwitchTheme.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using GeekAssistant.Forms;

internal static partial class GA_SetTheme {
    private static Timer HomeTheme_delayTimer = new() { Interval = 100 };

    private static PleaseWait pw = null;
    private static Preparing p = null;
    private static AppMode am = null;
    private static Donate d = null;
    private static Info i = null;
    private static Home h = null;
    private static Settings s = null;
    private static ToU t = null;
    public static void Run(Form f, bool initiating = false) {
        //Form Preparing = Application.OpenForms[0];
        //Form AppMode = new();
        //Form Home = new();
        //var form = f;
        //FormName = FormName[..(FormName.Length-2)]; //remove the () at the end which exists in C# form.Name
        initiatingbool = initiating;
        switch (f) { 
        case PleaseWait:
            pw = (PleaseWait)f;
            PleaseWait_Theme(); 
            break;
        case Preparing:
            p = (Preparing)f;
            p.Preparing_Label.ForeColor = c.colors.fg;
            p.BackColor = c.colors.bg;
            break;
        case AppMode:
            am = (AppMode)f;
            //Control am = AppMode;
            SetControlTheme(am);
            am.startup_dontShow.ForeColor = c.colors.fg;
            break;
        case Donate:
            d = (Donate)f;
            Donate_Theme();
            break;
        case Info:
            i = (Info)f;
            Info_Theme();
            break;
        case Home:
            h = (Home)f;
            HomeTheme_delayTimer.Tick += new(HomeTheme_delayTimer_Tick);
            if (c.S.VerboseLogging) {
                string ThemeString = "dark theme";
                if (c.S.DarkTheme)
                    ThemeString = "Light Theme";
                GA_SetProgressText.Run($"Switched to {ThemeString}.", -1);
            }
            if (!initiating & c.S.PerformAnimations) {
                if (c.S.DarkTheme) {
                    h.SwitchTheme_Back_UI.Top = -h.SwitchTheme_Back_UI.Height;
                    Animate.Run(h.SwitchTheme_Back_UI, "Top", new Home().Height, 1000);
                } else {
                    h.SwitchTheme_Back_UI.Top = new Home().Height;
                    Animate.Run(h.SwitchTheme_Back_UI, "Top", -h.SwitchTheme_Back_UI.Height, 1000);
                }
            }
            HomeTheme_delayTimer.Start();
            break;
        case Settings:
            s = (Settings)f;
            Settings_Theme();
            break;
        case ToU:
            t = (ToU)f;
            ToU_Theme();
            break;
        }

    }


    #region Donate
    private static void Donate_Theme() {
        //Donate d = new();
        SetControlsArrTheme(new Control[] { d, d.BitcoinAddress, d.BitcoinNote });
        {
            if (c.S.DarkTheme) {
                d.ButtonsBG_UI.BackColor = Color.FromArgb(32, 32, 32);
                d.Close_Button.FlatAppearance.MouseOverBackColor = Color.FromArgb(64, 64, 64);
                d.pic.Image = prop.x64.DonateHeart_Dark_64;
                d.title.ForeColor = Color.FromArgb(255, 191, 217);
                d.BitcoinAddressQR.Image = prop.xXX.BTCAddressQR_Dark;
                d.GooglePayLink.Image = prop.x16.linkIcon_dark_16;
            } else {
                d.ButtonsBG_UI.BackColor = Color.WhiteSmoke;
                d.Close_Button.FlatAppearance.MouseOverBackColor = Color.Silver;
                d.pic.Image = prop.x64.DonateHeart_64;
                d.title.ForeColor = Color.FromArgb(128, 0, 87);
                d.BitcoinAddressQR.Image = prop.xXX.BTCAddressQR;
                d.GooglePayLink.Image = prop.x16.linkIcon_16;
            }
            d.GeekAssistant_PictureBox.BackColor = d.ButtonsBG_UI.BackColor;
        }
    }

    #endregion

    #region Info
    private static void Info_Theme() {
        //Info i = new();
        SetControlsArrTheme(new Control[] { i, i.info_PictureBox, i.msg_Textbox, i.No_Button });
        {
            if (c.S.DarkTheme) {
                i.ButtonsBG_UI.BackColor = Color.FromArgb(32, 32, 32);
                i.Copy_Button.FlatAppearance.MouseOverBackColor = Color.FromArgb(64, 64, 64);
                i.Yes_Button.FlatAppearance.MouseOverBackColor = Color.FromArgb(64, 64, 64);
                i.No_Button.FlatAppearance.MouseOverBackColor = Color.FromArgb(64, 64, 64);
            } else {
                i.ButtonsBG_UI.BackColor = Color.WhiteSmoke;
                i.Copy_Button.FlatAppearance.MouseOverBackColor = Color.Silver;
                i.Yes_Button.FlatAppearance.MouseOverBackColor = Color.Silver;
                i.No_Button.FlatAppearance.MouseOverBackColor = Color.Silver;
            }

            i.Yes_Button.ForeColor = i.title_Label.ForeColor;
            i.Yes_Button.FlatAppearance.MouseDownBackColor = i.title_Label.ForeColor;
            i.GeekAssistant_PictureBox.BackColor = i.ButtonsBG_UI.BackColor;
        }
    }
    #endregion

    #region Home 
    private static bool initiatingbool = false;
    private static void HomeTheme_delayTimer_Tick(object sender, EventArgs e) {
        //Home h = new();
        Control[] Controls_array = new Control[] {
            h,
            h.log,
            h.ManualInfo_GroupBox,
            h.ProgressFakeBG_UI,
            h.Main_Tabs,
            h.PrepareYourDevice_Tab,
            h.FlashImg_Tab,
            h.MoreTools_Tab,
            h.BootloaderUnlockable_CheckBox,
            h.CustomROM_CheckBox,
            h.CustomRecovery_CheckBox
        };
        MetroFramework.Interfaces.IMetroControl[] MetroControls_array = new MetroFramework.Interfaces.IMetroControl[] {
            h.Main_Tabs,
            h.PrepareYourDevice_Tab,
            h.FlashImg_Tab,
            h.MoreTools_Tab,
            h.GA_About_Label,
            h.UnlockBL_Label,
            h.MagiskRoot_Label,
            h.Manufacturer_ComboBox,
            h.AndroidVersion_ComboBox,
            h.bar,
            h.UnlockBL_Button,
            h.MagiskRoot_Button,
            h.FlashZip_Button,
            h.ProgressBarLabel
        };//, c.Home.FlashZip_ChooseFile_TextBox, common.Home.manualCMD_TextBox };

        SetControlsArrTheme_Metro(MetroControls_array);

        if (initiatingbool) {

            for (int i = 0; i < Controls_array.Length; i++){
                Controls_array[i].BackColor = c.colors.bg;
                Controls_array[i].ForeColor = c.colors.fg;
            } 

        } else SetControlsArrTheme(Controls_array);

        var h_addb = h.AutoDetectDeviceInfo_Button;
        if (c.S.DarkTheme) {
            // Main.SwitchTheme_Mid_UI.BackColor = Color.FromArgb(235, 245, 235)
            // Transition.run(Main.SwitchTheme_Mid_UI, "BackColor", Color.FromArgb(0, 20, 0), New TransitionType_CriticalDamping(1000))
            h.Main_ToolTip.BackColor = SystemColors.InfoText;
            h.Main_ToolTip.ForeColor = SystemColors.Info;
            {
                {
                    h_addb.ForeColor = Color.FromArgb(95, 191, 119);
                    h_addb.BackColor = Color.FromArgb(64, 64, 64);
                    h_addb.FlatAppearance.MouseDownBackColor = Color.Honeydew;
                    h_addb.FlatAppearance.MouseOverBackColor = Color.FromArgb(74, 74, 74);
                    h_addb.Image = prop.x64.AutoDetect_dark_64;
                }
                // .DeviceState_Label.ForeColor = Color.FromArgb(95, 191, 119)
                h.Donate_Button.Icon = prop.x24.DonateHeart_Dark_24;
                h.SwitchTheme_Button.Icon = prop.x24.Theme_Dark_24;
                h.GeekAssistant.Image = prop.GA.Geek_Assistant___25_;
                h.Feedback_Button.Icon = prop.x24.Smile_dark_24;
                h.About_Button.Icon = prop.x24.ToU_dark_24;
                h.Settings_Button.Icon = prop.x24.Settings_dark_24;
                h.ShowLog_Button.Icon = prop.x24.Commands_dark_24;
                //hom.manualCMD_TextBox.Icon = prop.x16.Commands_dark_16;
            }
        } else // Light Theme
          {
            // Main.SwitchTheme_Mid_UI.BackColor = Color.FromArgb(0, 20, 0)
            // Transition.run(Main.SwitchTheme_Mid_UI, "BackColor", Color.FromArgb(235, 245, 235), New TransitionType_CriticalDamping(1000))
            h.Main_ToolTip.BackColor = SystemColors.Info;
            h.Main_ToolTip.ForeColor = SystemColors.InfoText;
            {
                {
                    h_addb.ForeColor = Color.FromArgb(0, 128, 32);
                    h_addb.BackColor = Color.Honeydew;
                    h_addb.FlatAppearance.MouseDownBackColor = Color.FromArgb(64, 64, 64);
                    h_addb.FlatAppearance.MouseOverBackColor = Color.FromArgb(230, 245, 230);
                    h_addb.Image = prop.x64.AutoDetect_64;
                }
                // .DeviceState_Label.ForeColor = Color.FromArgb(0, 128, 32)
                h.Donate_Button.Icon = prop.x24.DonateHeart_24;
                h.SwitchTheme_Button.Icon = prop.x24.Theme_Light_24;
                h.GeekAssistant.Image = prop.GA.Geek_Assistant;
                h.Feedback_Button.Icon = prop.x24.Smile_24;
                h.About_Button.Icon = prop.x24.ToU_24;
                h.Settings_Button.Icon = prop.x24.Settings_24;
                h.ShowLog_Button.Icon = prop.x24.Commands_24;
                //hom.manualCMD_TextBox.Icon = prop.x16.Commands_16;
            }
        }
        // Animate.Run(Main.log, "ForeColor", Current_fgColor())
        HomeTheme_delayTimer.Enabled = false;
    }
    #endregion

    #region PleaseWait
    private static void PleaseWait_Theme() {
        //PleaseWait p = new();
        SetControlsArrTheme(new Control[] { pw,  pw.PleaseWait_ProgressText });

        var p_spb = pw.StopProcess_Button;
        var p_spb_fa = p_spb.FlatAppearance;
        if (c.S.DarkTheme) {
            //p.PleaseWait_text.ForeColor = Color.FromArgb(0, 200, 0);
            {
                p_spb.ForeColor = Color.FromArgb(230, 255, 230);
                p_spb.BackColor = Color.FromArgb(55, 28, 25);
                {
                    p_spb_fa.MouseOverBackColor = Color.FromArgb(128, 0, 0);
                    p_spb_fa.MouseDownBackColor = Color.FromArgb(80, 0, 0);
                }
            }
        } else {
            pw.PleaseWait_text.ForeColor = Color.FromArgb(0, 100, 0);
            pw.StopProcess_Button.BackColor = Color.FromArgb(255, 228, 225);
            {
                p_spb.BackColor = Color.FromArgb(255, 228, 225);
                p_spb.ForeColor = Color.DarkRed;
                {
                    p_spb_fa.MouseOverBackColor = Color.FromArgb(240, 128, 128);
                    p_spb_fa.MouseDownBackColor = Color.FromArgb(128, 0, 0);
                }
            }
        }
    }
    #endregion

    #region Settings
    private static void Settings_Theme() {
        //Settings s = new();
        SetControlsArrTheme(new Control[] { s, s.ResetGA_GroupBox, s.Close_Button });

        var s_rsa = s.ResetGA_SelectAll;
        if (c.S.DarkTheme) {
            s.ButtonsBG_UI.BackColor = Color.FromArgb(32, 32, 32);
            s.ResetGA_LogsOnly_CheckBox_Label.ForeColor = Color.FromArgb(100, 100, 100);
            s.ResetGA_Settings_CheckBox_Label.ForeColor = Color.FromArgb(100, 100, 100);
            s.OpenLogsFolder.ForeColor = Color.FromArgb(100, 100, 100);
            s.ResetGA.FlatAppearance.MouseOverBackColor = Color.FromArgb(135, 83, 59);
            {
                s_rsa.Image = prop.x24.SelectAll_W_24;
                s_rsa.FlatAppearance.MouseOverBackColor = Color.FromArgb(64, 64, 64);
                s_rsa.FlatAppearance.MouseDownBackColor = Color.FromArgb(100, 100, 100);
            }

            s.SettingsIcon_PictureBox.Image = prop.x64.Settings_dark_64;
            s.SettingsTitle_Label.ForeColor = Color.FromArgb(184, 243, 254);
            s.Close_Button.FlatAppearance.MouseOverBackColor = Color.FromArgb(64, 64, 64);
        } else {
            s.ButtonsBG_UI.BackColor = Color.WhiteSmoke;
            s.ResetGA_LogsOnly_CheckBox_Label.ForeColor = Color.FromArgb(64, 64, 64);
            s.ResetGA_Settings_CheckBox_Label.ForeColor = Color.FromArgb(64, 64, 64);
            s.OpenLogsFolder.ForeColor = Color.FromArgb(64, 64, 64);
            s.ResetGA.FlatAppearance.MouseOverBackColor = Color.FromArgb(255, 203, 179);
            {
                s_rsa.Image = prop.x24.SelectAll_B_24;
                s_rsa.FlatAppearance.MouseOverBackColor = Color.Silver;
                s_rsa.FlatAppearance.MouseDownBackColor = Color.FromArgb(64, 64, 64);
            }
            s.SettingsIcon_PictureBox.Image = prop.x64.Settings_64;
            s.SettingsTitle_Label.ForeColor = Color.FromArgb(0, 106, 128);
            s.Close_Button.FlatAppearance.MouseOverBackColor = Color.Silver;

        }
        {
            s.ResetGA.BackColor = s.ButtonsBG_UI.BackColor;
            s.ResetGA_SelectAll.BackColor = s.ButtonsBG_UI.BackColor;
            s.GeekAssistant_PictureBox.BackColor = s.ButtonsBG_UI.BackColor;
            s.Close_Button.BackColor = s.BackColor;
        }
    }
    #endregion

    #region ToU
    private static void ToU_Theme() {
        //ToU t = new();
        SetControlsArrTheme(new Control[] { t, t.TermsOfUse_Box, t.ToU_Reject, t.ToU_Accept });

        // Do this hell
        if (c.S.DarkTheme) {
            t.ButtonsBG_UI.BackColor = Color.FromArgb(32, 32, 32);
            t.Icon_PictureBox.Image = prop.x64.ToU_dark_64;
            t.ToUTitle_Label.ForeColor = Color.FromArgb(178, 255, 220);
        } else {
            t.ButtonsBG_UI.BackColor = Color.WhiteSmoke;
            t.Icon_PictureBox.Image = prop.x64.ToU_64;
            t.ToUTitle_Label.ForeColor = Color.FromArgb(0, 102, 71);
        }

        t.AcceptCheck_ToU.ForeColor = t.ToUTitle_Label.ForeColor;
        t.Copyright_Label.ForeColor = t.ToUTitle_Label.ForeColor;
        t.AcceptCheck_ToU.ForeColor = c.colors.fg;
        t.DontShow_ToU.ForeColor = c.colors.fg;
        t.AcceptCheck_ToU.BackColor = t.ButtonsBG_UI.BackColor;
        t.DontShow_ToU.BackColor = t.ButtonsBG_UI.BackColor;
        t.GeekAssistant_PictureBox.BackColor = t.ButtonsBG_UI.BackColor;
    }
    #endregion


    #region Theming Mechanism

    private static void SetControlsArrTheme(Control[] ControlsObj) {
        for (int i = 0; i < ControlsObj.Length; i++)
             SetControlTheme(  ControlsObj[i]); 
    } 
    private static void SetControlTheme(  Control ControlObj) {
        Animate.Run(  ControlObj, "ForeColor", c.colors.fg);
        Animate.Run(  ControlObj, "BackColor", c.colors.bg);
    }

    private static void SetControlsArrTheme_Metro(MetroFramework.Interfaces.IMetroControl[] ControlsObj) {
        for (int i = 0; i < ControlsObj.Length; i++)
            SetControlTheme_Metro(  ControlsObj[i]);
    }
    private static void SetControlTheme_Metro(  MetroFramework.Interfaces.IMetroControl ControlObj) {
        ControlObj.Theme = c.colors.Metro;
        // Cannot animate "Theme" 'Transition.run(ControlObj, "Theme", Current_Theme_Metro(), New TransitionType_CriticalDamping(500))
    }

    #endregion
}